dali/pipeline/graph_descr.cc

```diff
@@ -71,358 +73,444 @@ void CheckOpConstraints(const OpSpec &spec) {
     meta.is_support = spec.IsArgumentInput(i);
-    meta.is_cpu = spec.InputDevice(i) == "cpu" ? true : false;
+    meta.storage_device = DeviceStringToTensorStorage(spec.InputDevice(i));
 
-    vector<TensorMeta> &consumer_info = tensor_consumers_[spec.Input(i)];
-    consumer_info.push_back(meta);
+    // Insert new tensor consumer
+    tensor_nodes_[consumed_tensor_id].consumer_edges.push_back(meta);
   }
 
   // Mark this op as the source of its output tensors
+  // Create TensorNodes for outputs and respective edges for this OpNode -> TensorNodes
   for (int i = 0; i < spec.NumOutput(); ++i) {
-    string name = spec.Output(i);
-
     // Set the producer info for this tensor
     TensorMeta meta;
-    meta.node = new_node->id;
+    meta.node = new_node.id;
     meta.index = i;
     meta.is_support = spec.GetArgument<string>("device") == "support";
-    meta.is_cpu = spec.OutputDevice(i) == "cpu" ? true : false;
+    meta.storage_device = DeviceStringToTensorStorage(spec.OutputDevice(i));
+
+    string name = spec.Output(i);
 
-    auto ret = tensor_producers_.insert({name, meta});
-    DALI_ENFORCE(ret.second, "Operator '" + spec.name() +
+    // Place new Tensor with producer info and add edge to OpNode.
+    auto &new_tensor = PlaceNewTensor();
+    meta.tensor = new_tensor.id;
+    new_tensor.producer_edge = meta;
+    new_tensor.name = name;
+    new_node.children_tensors.push_back(new_tensor.id);
+
+    auto it_inserted = tensor_name_to_id_.insert({name, new_tensor.id});
+    DALI_ENFORCE(it_inserted.second, "Operator '" + spec.name() +
         "' has output with name " + name + ", but output "
         "with this name already exists as output of op '" +
-        this->node(TensorSourceID(name)).spec.name() + "'");
+        this->Node(TensorSourceID(name)).spec.name() + "'");
   }
 }
 
 void OpGraph::InstantiateOperators() {
   // traverse devices by topological order (support, cpu, mixed, gpu)
-  for (auto &node : support_nodes_) {
-    if (node.op)
-      continue;
-    auto &spec = node.spec;
-    string device = spec.GetArgument<string>("device");
-    node.op = SupportOperatorRegistry::Registry().Create(spec.name(), spec, &device);
-  }
-  for (auto &node : cpu_nodes_) {
-    if (node.op)
-      continue;
-    auto &spec = node.spec;
-    string device = spec.GetArgument<string>("device");
-    node.op = CPUOperatorRegistry::Registry().Create(spec.name(), spec, &device);
-  }
-  for (auto &node : mixed_nodes_) {
-    if (node.op)
-      continue;
-    auto &spec = node.spec;
-    string device = spec.GetArgument<string>("device");
-    node.op = MixedOperatorRegistry::Registry().Create(spec.name(), spec, &device);
-  }
-  for (auto &node : gpu_nodes_) {
-    if (node.op)
-      continue;
-    auto &spec = node.spec;
-    string device = spec.GetArgument<string>("device");
-    node.op = GPUOperatorRegistry::Registry().Create(spec.name(), spec, &device);
+  DALIOpType order[] = { DALIOpType::DALI_SUPPORT, DALIOpType::DALI_CPU,
```

Comment:
I wonder if this order is really necessary? Do we have global topological order? If we do, isn't it  sufficient?

Reply:
I didn't dig deep enough into it, so I just go in the order of stages.